### PR TITLE
DOT-1463 delete group

### PR DESCRIPTION
**Pull request by edwh**
Add client and server function to allow Administrators to delete groups.

This is on top of the as-yet-unreleased DOT-810, for testing reasons.

### CODE REVIEW

File: app/Group.php [hunk 1]
```diff
@@ -243,6 +244,27 @@ public function getLocation()
         return rtrim($this->location);
     }
 
+    public function canDelete() {
```
Comment (ngm) [inline app/Group.php:247]: Minor one - not for now but for retro - PHP code guidelines that I've seen (PSR-2, Laravel's own) tend to go with opening brace on new line.

File: app/Http/Controllers/GroupController.php [hunk 2]
```diff
@@ -766,22 +790,55 @@ public function edit(Request $request, $id)
...
-    // TODO: is this alive?
     public function delete($id)
     {
-        if (FixometerHelper::hasRole($this->user, 'Administrator')) {
-            $r = $this->Group->delete($id);
+        $group = Group::where('idgroups', $id)->first();
+
+        $name = $group->name;
+
+        if (FixometerHelper::hasRole(Auth::user(), 'Administrator') && $group->canDelete()) {
```
Comment (ngm) [inline app/Http/Controllers/GroupController.php:799]: Not big deal here, but we have a method on user called hasRole which would be better to use in general.  I feel like FixometerHelper class in general should be deprecated over time.

File: app/Http/Controllers/GroupController.php [hunk 3]
```diff
@@ -766,22 +790,55 @@ public function edit(Request $request, $id)
...
+            $r = $group->delete($id);
+            $r = true;
             if ( ! $r) {
-                $response = 'd:err';
+                return redirect('/user/forbidden');
```
Comment (ngm) [inline app/Http/Controllers/GroupController.php:803]: Does this route still exist?

File: resources/js/components/GroupActions.vue [hunk 4]
```diff
@@ -67,6 +91,12 @@ export default {
       })
 
       this.$emit('left')
+    },
+    deleteGroup() {
+      this.$refs.confirmDelete.show()
+    },
+    async deleteConfirmed() {
+      window.location = '/group/delete/' + this.idgroups
```
Comment (ngm) [inline resources/js/components/GroupActions.vue:99]: Is it OK that the delete action is a GET?  As it is behind permission checks, probably OK.  Possibly a more philosophical point, could discuss at retro.

File: tests/Feature/Groups/GroupDeleteTest.php [hunk 5]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace Tests\Feature\Groups;
+
+use App\Group;
+use App\Role;
+use Tests\TestCase;
+
+class GroupDeleteTest extends TestCase {
```
Comment (ngm) [inline tests/Feature/Groups/GroupDeleteTest.php:9]: Should we also have a test for the check regarding not being able to delete if past events / devices exist?

**Review by ngm — APPROVED** [review]
LGTM :+1: Some minor comments, no blockers though.  Could add in an extra test if time permits.

File: app/Group.php (hunk 1 above)
Comment (edwh) [reply app/Group.php:247]: I've changed my IDE.

File: app/Http/Controllers/GroupController.php (hunk 2 above)
Comment (edwh) [reply app/Http/Controllers/GroupController.php:799]: Thanks.  Changed.

File: app/Http/Controllers/GroupController.php (hunk 3 above)
Comment (edwh) [reply app/Http/Controllers/GroupController.php:803]: Yes - see https://restarters.net/user/forbidden

File: resources/js/components/GroupActions.vue (hunk 4 above)
Comment (edwh) [reply resources/js/components/GroupActions.vue:99]: No, but our use of HTTP verbs is all over the place.  Added to retro.

File: tests/Feature/Groups/GroupDeleteTest.php (hunk 5 above)
Comment (edwh) [reply tests/Feature/Groups/GroupDeleteTest.php:9]: Yes, and adding that finds bugs.  Tests do that.  Can you rereview?

**Review by ngm — APPROVED** [review]
Great, changes all look good now.

**Pull request merged**